peer_heap.js

```diff
@@ -20,89 +20,88 @@
 }
 
-PeerHeap.prototype.choose = function choose(threshold, filter) {
-    var self = this;
-
-    if (!self.array.length) {
-        return null;
-    }
-
-    var el;
-    if (filter) {
-        el = self._chooseFilteredEl(threshold, filter);
-    } else {
-        el = self._chooseEl(threshold);
-    }
-
-    if (!el) {
-        return null;
-    }
-
-    return el.peer;
-
-};
-
-PeerHeap.prototype._chooseEl = function _chooseEl(threshold) {
+PeerHeap.prototype.chooseWeightedRandom = function chooseWeightedRandom(threshold, filter) {
     var self = this;
 
-    var el = self.array[0];
-    if (el.score <= threshold) { // TODO: why inclusive?
-        return null;
-    }
-
-    return el;
-};
+    // We select into this set as long as we have ranges that overlap. Then,
+    // we introduce randomness into the set and choose
+    var seedPeer = self.array[0].peer;
+    var seedRange = seedPeer.pendingWeightedRange();
+    var set = [seedPeer];
+    var minRangeStart = seedRange[0];
+    var i;
 
-PeerHeap.prototype._chooseFilteredEl = function _chooseFilteredEl(threshold, filter) {
-    var self = this;
-
-    // TODO: is it worth it to unroll the first iteration of the loop below so
-    // that we incur minimal overhead for "the top of heap is okay" case?
-
-    // NOTE: assumes self._stack starts off empty
     self._stack.push(0);
-    while (self._stack.length) {
-        var i = self._stack.shift();
 
+    while (self._stack.length) {
+        i = self._stack.shift();
         var el = self.array[i];
-        if (el.score <= threshold) { // TODO: why inclusive?
-            break;
-        }
 
-        if (!filter || filter(el.peer)) {
-            return el;
+        var range = el.peer.pendingWeightedRange();
+
+        if (range[1] <= minRangeStart) {
+            // If this range ends before the range with the smallest start
+            // begins, then it's no longer overlapping. Continue without
+            // adding the left and right children to the search stack.
+            continue;
+        } else if (!filter || filter(el.peer)) {
+            minRangeStart = Math.min(minRangeStart, range[0]);
+            set.push(el.peer);
         }
 
+        // Continue DFS down heap
         var left = 2 * i + 1;
+        var right = 2 * i + 2;
         if (left < self.array.length) {
-            var right = left + 1;
+            self._stack.push(left);
             if (right < self.array.length) {
-                if (self.array[right].score > self.array[left].score) {
-                    self._stack.push(right, left);
-                } else {
-                    self._stack.push(left, left);
-                }
-            } else {
-                self._stack.push(left);
+                self._stack.push(right);
             }
         }
     }
+
     self._stack.length = 0;
 
-    return null;
+    var chosenPeer = null;
+    var highestProbability = 0;
+    var probability;
+    for (i = 0; i < set.length; i++) {
```

Comment:
you could inline this into the loop above without any need to keep the temporary set; doing so may well reduce the overhead of this method.